### PR TITLE
Throw exception if instantiating IOSDevice on non-mac os platform

### LINKED ISSUE #36008: throw or fail assertion if `IosDevice` constructor called on non Mac OS platform
Currently, there is a check in the `IosDevice` constructor if `platform.isMacOS` [here](https://github.com/flutter/flutter/blob/735352a8c0a462b234a20ee6ed50bc03d4a20079/packages/flutter_tools/lib/src/ios/devices.dart#L184), and prints an error if it is not. This constructor should never get called if not on Mac OS, so there should be either a thrown exception or an assert instead.

Should be fixed after https://github.com/flutter/flutter/pull/34669 lands, as this code was refactored there.

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

Add an assert to verify we are not instantiating `IOSDevice` class on non-Mac OS platform. Doing this uncovered that most of the rest of the tests in this file were actually going down this path, so I fixed that too.

## Related Issues

fixes [#36008](https://github.com/flutter/flutter/issues/36008)
## Tests

I added the following tests:

I added a new test to verify the assert error on non-mac, and then fixed the many of the other tests to be on mac platform.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change (shouldn't be).

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
Are we sure that we don't end up going down this code path on windows/linux machines?

File: packages/flutter_tools/test/general.shard/ios/devices_test.dart [hunk 1]
```diff
@@ -46,6 +47,26 @@ void main() {
       IMobileDevice: () => mockIMobileDevice,
     });
 
+    testUsingContext('throws assertion error if platform is not mac os', () async {
+      when(mockIMobileDevice.isInstalled).thenReturn(true);
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/ios/devices_test.dart]: Would imobiledevice be installed on Linux? Is there some series of actions that a user could take to get into this state?

Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/ios/devices_test.dart]: @zanderso @jonahwilliams I can say with confidence that within the tool code the only place this gets instantiated is in the `IOSDevices` class, which has `bool get supportsPlatform => platform.isMacOS;`. It's possible there's some other tool implementation that instantiates it, which is why I split this out into its own PR.

Looking at the origin of the platform check, it goes back over 3 years ([here](https://github.com/flutter/flutter/commit/67046f933dce5cb8713ae673a664ee323a267840#diff-4e6a9727ba10772d640b597dbdd61d03R105).

Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/ios/devices_test.dart]: good point, I need to re-write this test.

**Comment by christopherfujino** [comment]
> Are we sure that we don't end up going down this code path on windows/linux machines?

I can say with confidence that within the tool code the only place this gets instantiated is in the IOSDevices class, which has bool get supportsPlatform => platform.isMacOS;. It's possible there's some other tool implementation that instantiates it, which is why I split this out into its own PR.

Looking at the origin of the platform check, it goes back over 3 years (here).

File: packages/flutter_tools/test/general.shard/ios/devices_test.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/ios/devices_test.dart]: Done

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 2]
```diff
@@ -125,10 +125,7 @@ class IOSDevice extends Device {
           ephemeral: true,
       ) {
     if (!platform.isMacOS) {
-      printError('Cannot control iOS devices or simulators. ideviceinstaller and iproxy are not available on your platform.');
-      _installerPath = null;
-      _iproxyPath = null;
-      return;
+      throw UnsupportedError('Control of iOS devices or simulators only supported on Mac OS.');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Sorry for only now thinking about this a bit more, but I wonder if we should instead do the following:
1. Initialize `_installerPath` and `_iproxyPath` only on the first access.
2. throwing the UnsupportedError at the very top of getAttachedDevices().

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: I can write getters that only initialize the variables when called. Curious, what's the advantage of moving the platform check later? To allow instantiating the class on non mac platforms?

**Review by zanderso — APPROVED** [review]
lgtm

**Pull request merged**